# iNaturalist status updates by state

Using the file produced in the collate-status-taxa.ipynb: `inat-aust-status-taxa.csv`, generate lists to update iNaturalist statuses

**Next steps:**
State by state establish the changes that need to be made:
    a. new - any new species that appear in the state lists but do not have a status in inaturalist (new template)
    b. updates - any changes to information which was added by us previously (user_id = 708886) (update template, action='UPDATE')
    c. removals - any statuses which were added by us previously (user_id = 708886) list which are incorrect (update template, action='REMOVE')
    d. flags - are there any statuses by other users that need to be flagged?

## Prep - common to all states
1. Read in the inaturalist statuses & filter out this state
2. Read in the inaturalist taxa list
3. Read in the state sensitive list
4. Attempt to match the state statuses to an IUCN equivalent


### 1. iNaturalist statuses

In [51]:
import pandas as pd

projectdir = "/Users/oco115/PycharmProjects/authoritative-lists/" # basedir for this gh project
# projectdir = "/Users/new330/IdeaProjects/authoritative-lists/" # basedir for this gh project
sourcedir = projectdir + "source-data/inaturalist-statuses/"
listdir = projectdir + "current-lists/"


# read in the statuses
taxastatus = pd.read_csv(sourcedir + "inat-aust-status-taxa.csv", encoding='UTF-8',na_filter=False,dtype=str) ## Read inaturalist conservation statuses file
taxastatus.head(3)

id taxon_id  user_id place_id source_id  \
0  166449    38493  1138587     7830             
1  234788   918383   702203     9994             
2  234789   918383   702203     7308             

                            authority status  \
0  Flora and Fauna Guarantee Act 1988     CR   
1           Atlas of Living Australia    NT    
2           Atlas of Living Australia     LC   

                                                 url description geoprivacy  \
0                                                                  obscured   
1  https://bie.ala.org.au/species/https://id.biod...                          
2  https://bie.ala.org.au/species/https://id.biod...                          

   ...         genus specificEpithet infraspecificEpithet  \
0  ...     Eulamprus      kosciuskoi                        
1  ...  Chiloschista     phyllorhiza                        
2  ...  Chiloschista     phyllorhiza                        

               modified            scientificName taxonRank  \
0  2021-03-01T10:35:01Z      Eulamprus kosciuskoi   species   
1  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   
2  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   

                                          references preferred_common_name  \
0  http://reptile-database.reptarium.cz/search.ph...                         
1  http://www.catalogueoflife.org/annual-checklis...                         
2  http://www.catalogueoflife.org/annual-checklis...                         

  is_active current_synonymous_taxon_ids  
0                                         
1                                         
2                                         

[3 rows x 34 columns]

In [52]:
def filter_state_statuses(stateregex: str, urlregex: str):
    authoritydf = taxastatus['authority'].drop_duplicates().sort_values()
    authoritydf = authoritydf[pd.Series(authoritydf).str.contains(stateregex)]
    # print(authoritydf)
    urldf = taxastatus['url'].drop_duplicates().sort_values()
    urldf = urldf[pd.Series(urldf).str.contains(urlregex)]
    placedisplaydf = taxastatus['place_display_name'].drop_duplicates().sort_values()
    placedisplaydf = placedisplaydf[pd.Series(placedisplaydf).str.contains(stateregex)]
    placedf = taxastatus['place_name'].drop_duplicates().sort_values()
    placedf = placedf[pd.Series(placedf).str.contains(stateregex)]
    # concat all and remove duplicates
    statedf = pd.concat([taxastatus.apply(lambda row: row[taxastatus['place_display_name'].isin(placedisplaydf)]),
                         taxastatus.apply(lambda row: row[taxastatus['place_name'].isin(placedf)]),
                         # taxastatus.apply(lambda row: row[taxastatus['url'].isin(urldf)]),
                         taxastatus.apply(
                             lambda row: row[taxastatus['authority'].isin(authoritydf)])]).drop_duplicates()
    return statedf.sort_values(['taxon_id', 'user_id'])
inatstatuses = filter_state_statuses("Northern Territory|NT NRETAS", " ")
inatstatuses.rename(columns={'id':'status_id','id_y':'taxon_id_y'},inplace=True)
inatstatuses

status_id taxon_id user_id place_id source_id                  authority  \
1001    162724  1134239  702203     9994                   Northern Territory   
927     165448    12647  702203     9994                   Northern Territory   
289     263901  1289379  702203     9994            Atlas of Living Australia   
1251    152435    19250  708886     9994     16652                  NT NRETAS   
215     170163    20166  702203     9994            Atlas of Living Australia   
1249    152433    38633  708886     9994     16652                  NT NRETAS   
1252    152436    40743  708886     9994     16652                  NT NRETAS   
1248    152432    41326  708886     9994     16652                  NT NRETAS   
986     139906   698942             9994                                        
1247    152431    73180  708886     9994     16652                  NT NRETAS   
1250    152434    74425  708886     9994     16652                  NT NRETAS   
1       234788   918383  702203     9994            Atlas of Living Australia   

             status                                                url  \
1001             LC  https://bie.ala.org.au/species/https://id.biod...   
927              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
289              VU  https://bie.ala.org.au/species/urn:lsid:biodiv...   
1251     endangered  https://lists.ala.org.au/speciesListItem/list/...   
215              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
1249     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1252     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1248     endangered  https://lists.ala.org.au/speciesListItem/list/...   
986   least concern  https://bie.ala.org.au/species/http://id.biodi...   
1247     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1250     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1               NT   https://bie.ala.org.au/species/https://id.biod...   

                          description geoprivacy  ...         genus  \
1001                                        open  ...    Cryptandra   
927                                         open  ...    Epthianura   
289                                         open  ...      Chloebia   
1251                                    obscured  ...     Polytelis   
215                                               ...         Ninox   
1249                                    obscured  ...   Bellatorias   
1252                                    obscured  ...                 
1248                                    obscured  ...    Macroderma   
986   Atlas of Living Australia (ALA)             ...      Duboisia   
1247                                     private  ...     Pezoporus   
1250                                    obscured  ...  Hipposideros   
1                                                 ...  Chiloschista   

     specificEpithet infraspecificEpithet              modified  \
1001         gemmata                       2020-09-25T18:01:37Z   
927           crocea                       2021-09-17T08:46:17Z   
289         gouldiae                       2022-10-20T02:55:37Z   
1251      alexandrae                       2019-08-27T01:09:01Z   
215        connivens                       2021-07-28T02:17:03Z   
1249           obiri                       2019-04-30T15:19:21Z   
1252                                                              
1248           gigas                       2019-08-27T01:58:05Z   
986        hopwoodii                       2019-02-16T13:19:18Z   
1247    occidentalis                       2019-08-27T01:09:27Z   
1250       inornatus                       2019-01-05T10:25:06Z   
1        phyllorhiza                       2022-01-08T03:30:36Z   

                scientificName taxonRank  \
1001        Cryptandra gemmata   species   
927          Epthianura crocea   species   
289          Chloebia gouldiae   species   
1251      Polyte

### 2. iNaturalist taxonomy

In [53]:
# Output files contain these fields
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# so we need to match species from the state lists to the inat taxa to get the taxon_id

import zipfile
url = "https://www.inaturalist.org/taxa/inaturalist-taxonomy.dwca.zip"
filename = url.split("/")[-1]

z=zipfile.ZipFile(sourcedir + filename)

with z.open('taxa.csv') as from_archive:
    inattaxa = pd.read_csv(from_archive,dtype=str)
z.close()
inattaxa.head(3)


id                             taxonID                          identifier  \
0  1  https://www.inaturalist.org/taxa/1  https://www.inaturalist.org/taxa/1   
1  2  https://www.inaturalist.org/taxa/2  https://www.inaturalist.org/taxa/2   
2  3  https://www.inaturalist.org/taxa/3  https://www.inaturalist.org/taxa/3   

                         parentNameUsageID   kingdom    phylum class order  \
0   https://www.inaturalist.org/taxa/48460  Animalia       NaN   NaN   NaN   
1       https://www.inaturalist.org/taxa/1  Animalia  Chordata   NaN   NaN   
2  https://www.inaturalist.org/taxa/355675  Animalia  Chordata  Aves   NaN   

  family genus specificEpithet infraspecificEpithet              modified  \
0    NaN   NaN             NaN                  NaN  2021-11-02T06:05:44Z   
1    NaN   NaN             NaN                  NaN  2021-11-23T00:40:18Z   
2    NaN   NaN             NaN                  NaN  2022-12-27T07:33:16Z   

  scientificName taxonRank                                         references  
0       Animalia   kingdom  http://www.catalogueoflife.org/annual-checklis...  
1       Chordata    phylum  http://www.catalogueoflife.org/annual-checklis...  
2           Aves     class  http://www.catalogueoflife.org/annual-checklis...

### 3. State lists
Sensitive list: `geoprivacy` = `obscured`
overrides

In [54]:
# NT - No sensitive list

# sensitivelist = pd.read_csv(listdir + "sensitive-lists/NT-sensitive.csv")  # Qld sensitive list
# sensitivelist['scientificName'] = sensitivelist['scientificname'].str.replace('subsp. ', '', regex=False)
# sensitivelist['bionet_geoprivacy'] = 'obscured'
# # sensitivelist = sensitivelist.rename(columns={'taxonID':'wildnetTaxonID'})
# sensitivelist

In [55]:
conservationlist = pd.read_csv(listdir + "conservation-lists/NT-conservation.csv")  # NT sensitive list
conservationlist['scientificName'] = conservationlist['scientificName'].str.replace('subsp. ', '', regex=False)
# conservationlist = conservationlist.rename(columns={'taxonID':'bionetTaxonID'})
# conservationlist['bionet_geoprivacy'] = conservationlist['sensitivityClass'].apply(gpcondition)
conservationlist['bionet_geoprivacy'] = 'open'
conservationlist

family                 status                sourceStatus  \
0         NaN            Vulnerable   Vulnerable (extinct in NT)   
1         NaN             Vulnerable  Vulnerable (extinct in NT)   
2         NaN             Vulnerable  Vulnerable (extinct in NT)   
3         NaN             Vulnerable  Vulnerable (extinct in NT)   
4    Fabaceae             Vulnerable                  Vulnerable   
..        ...                    ...                         ...   
199       NaN  Critically Endangered       Critically Endangered   
200       NaN  Critically Endangered       Critically Endangered   
201       NaN  Critically Endangered       Critically Endangered   
202       NaN  Critically Endangered       Critically Endangered   
203       NaN  Critically Endangered       Critically Endangered   

                        vernacularName                 scientificName  \
0                              Kowari              Dasyuroides byrnei   
1               Red-tailed phascogale               Phascogale calura   
2                     Shark Bay mouse                Pseudomys fieldi   
3                       Western quoll              Dasyurus geoffroii   
4                        Latz's wattle                  Acacia latzii   
..                                 ...                            ...   
199                        Malleefowl                 Leipoa ocellata   
200                    Northern quoll             Dasyurus hallucatus   
201                   Plains-wanderer            Pedionomus torquatus   
202            Saddle Creek rocksnail          Vincentrachia desmonda   
203  Thick-billed grasswren (western)   Amytornis modestus indulkanna   

     taxonRemarks bionet_geoprivacy  
0          Mammal              open  
1          Mammal              open  
2          Mammal              open  
3          Mammal              open  
4             NaN              open  
..            ...               ...  
199          Bird              open  
200        Mammal              open  
201          Bird              open  
202  Invertebrate              open  
203          Bird              open  

[204 rows x 7 columns]

In [56]:
# join them in a way that works for inat (eg sensitive list, geoprivacy = 'obscured'
statelist =  conservationlist.drop_duplicates()
statelist

family                 status                sourceStatus  \
0         NaN            Vulnerable   Vulnerable (extinct in NT)   
1         NaN             Vulnerable  Vulnerable (extinct in NT)   
2         NaN             Vulnerable  Vulnerable (extinct in NT)   
3         NaN             Vulnerable  Vulnerable (extinct in NT)   
4    Fabaceae             Vulnerable                  Vulnerable   
..        ...                    ...                         ...   
199       NaN  Critically Endangered       Critically Endangered   
200       NaN  Critically Endangered       Critically Endangered   
201       NaN  Critically Endangered       Critically Endangered   
202       NaN  Critically Endangered       Critically Endangered   
203       NaN  Critically Endangered       Critically Endangered   

                        vernacularName                 scientificName  \
0                              Kowari              Dasyuroides byrnei   
1               Red-tailed phascogale               Phascogale calura   
2                     Shark Bay mouse                Pseudomys fieldi   
3                       Western quoll              Dasyurus geoffroii   
4                        Latz's wattle                  Acacia latzii   
..                                 ...                            ...   
199                        Malleefowl                 Leipoa ocellata   
200                    Northern quoll             Dasyurus hallucatus   
201                   Plains-wanderer            Pedionomus torquatus   
202            Saddle Creek rocksnail          Vincentrachia desmonda   
203  Thick-billed grasswren (western)   Amytornis modestus indulkanna   

     taxonRemarks bionet_geoprivacy  
0          Mammal              open  
1          Mammal              open  
2          Mammal              open  
3          Mammal              open  
4             NaN              open  
..            ...               ...  
199          Bird              open  
200        Mammal              open  
201          Bird              open  
202  Invertebrate              open  
203          Bird              open  

[204 rows x 7 columns]

### 4. Equivalent IUCN statuses

In [57]:
iucn_statuses = {'Not Evaluated', 'Data Deficient', 'Least Concern', 'Near Threatened', 'Vulnerable', 'Endangered', 'Critically Endangered', 'Extinct in the Wild' and 'Extinct'}
# sensitivelist.groupby(['status'])['status'].count()
statelist.groupby(['status'])['status'].count()

status
Critically Endangered     20
Endangered                52
Extinct                   11
Vulnerable               120
Vulnerable                 1
Name: status, dtype: int64

In [58]:
iucnStatusMappings = {
    'Vulnernable':'Vulnerable'
}

### 5. Determine best place ID to use

In [59]:
inatstatuses.groupby(['place_id','place_name','place_display_name'])['place_id'].count()
# looks like 6825

place_id  place_name          place_display_name    
9994      Northern Territory  Northern Territory, AU    12
Name: place_id, dtype: int64

## Merge iNaturalist statuses with State sensitive list on scientificName

1. Match - updates, even if the statuses are the same we'll update the links and values anyway
2. No match - statuses to be added (additions)
   1.1 No match and no taxnomy - search for synonyms
   1.2 No match
3. Merge the other direction to see if there are deletes?


In [60]:
# join to see which lists already have a status in inaturalist based on scientificName
mergedstatuses = statelist[['scientificName','status','bionet_geoprivacy']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='scientificName',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['scientificName'])
mergedstatuses


scientificName                 status bionet_geoprivacy status_id  \
135  Abrodictyum obscurum             Endangered              open       NaN   
185  Acacia equisetifolia  Critically Endangered              open       NaN   
4           Acacia latzii             Vulnerable              open       NaN   
136          Acacia peuce             Endangered              open       NaN   
5     Acacia praetermissa             Vulnerable              open       NaN   
..                    ...                    ...               ...       ...   
155    Xylopia monosperma             Endangered              open       NaN   
55        Zeuxine oblonga             Vulnerable              open       NaN   
58          Zyzomys maini             Vulnerable              open       NaN   
166     Zyzomys palatalis             Endangered              open       NaN   
191  Zyzomys pedunculatus  Critically Endangered              open       NaN   

    taxon_id user_id description iucn authority status_inat geoprivacy  \
135      NaN     NaN         NaN  NaN       NaN         NaN        NaN   
185      NaN     NaN         NaN  NaN       NaN         NaN        NaN   
4        NaN     NaN         NaN  NaN       NaN         NaN        NaN   
136      NaN     NaN         NaN  NaN       NaN         NaN        NaN   
5        NaN     NaN         NaN  NaN       NaN         NaN        NaN   
..       ...     ...         ...  ...       ...         ...        ...   
155      NaN     NaN         NaN  NaN       NaN         NaN        NaN   
55       NaN     NaN         NaN  NaN       NaN         NaN        NaN   
58       NaN     NaN         NaN  NaN       NaN         NaN        NaN   
166      NaN     NaN         NaN  NaN       NaN         NaN        NaN   
191      NaN     NaN         NaN  NaN       NaN         NaN        NaN   

    place_id place_display_name  
135      NaN                NaN  
185      NaN                NaN  
4        NaN                NaN  
136      NaN                NaN  
5        NaN                NaN  
..       ...                ...  
155      NaN                NaN  
55       NaN                NaN  
58       NaN                NaN  
166      NaN                NaN  
191      NaN                NaN  

[204 rows x 13 columns]

In [61]:
# prepare the export fields, common to New template and Update template
# new statuses
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username
mergedstatuses['new_authority'] = " Territory Parks and Wildlife Conservation Act 1976"
mergedstatuses['new_description'] = "Listed as Threatened - refer to https://nt.gov.au/environment"
mergedstatuses['new_url'] = "https://nt.gov.au/environment"
mergedstatuses['new_geoprivacy'] = "obscured"
mergedstatuses['new_place_id'] = '9994'  # Northern Territory
mergedstatuses['new_username'] = 'peggydnew'
mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
mergedstatuses['new_status'] = mergedstatuses['status'].fillna('Threatened')
mergedstatuses

scientificName                 status bionet_geoprivacy status_id  \
135  Abrodictyum obscurum             Endangered              open       NaN   
185  Acacia equisetifolia  Critically Endangered              open       NaN   
4           Acacia latzii             Vulnerable              open       NaN   
136          Acacia peuce             Endangered              open       NaN   
5     Acacia praetermissa             Vulnerable              open       NaN   
..                    ...                    ...               ...       ...   
155    Xylopia monosperma             Endangered              open       NaN   
55        Zeuxine oblonga             Vulnerable              open       NaN   
58          Zyzomys maini             Vulnerable              open       NaN   
166     Zyzomys palatalis             Endangered              open       NaN   
191  Zyzomys pedunculatus  Critically Endangered              open       NaN   

    taxon_id user_id description iucn authority status_inat  ... place_id  \
135      NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
185      NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
4        NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
136      NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
5        NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
..       ...     ...         ...  ...       ...         ...  ...      ...   
155      NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
55       NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
58       NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
166      NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
191      NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   

    place_display_name                                      new_authority  \
135                NaN   Territory Parks and Wildlife Conservation Act...   
185                NaN   Territory Parks and Wildlife Conservation Act...   
4                  NaN   Territory Parks and Wildlife Conservation Act...   
136                NaN   Territory Parks and Wildlife Conservation Act...   
5                  NaN   Territory Parks and Wildlife Conservation Act...   
..                 ...                                                ...   
155                NaN   Territory Parks and Wildlife Conservation Act...   
55                 NaN   Territory Parks and Wildlife Conservation Act...   
58                 NaN   Territory Parks and Wildlife Conservation Act...   
166                NaN   Territory Parks and Wildlife Conservation Act...   
191                NaN   Territory Parks and Wildlife Conservation Act...   

                                       new_description  \
135  Listed as Threatened - refer to https://nt.gov...   
185  Listed as Threatened - refer to https://nt.gov...   
4    Listed as Threatened - refer to https://nt.gov...   
136  Listed as Threatened - refer to https://nt.gov...   
5    Listed as Threatened - refer to https://nt.gov...   
..                                                 ...   
155  Listed as Threatened - refer to https://nt.gov...   
55   Listed as Threatened - refer to https://nt.gov...   
58   Listed as Threatened - refer to https://nt.gov...   
166  Listed as Threatened - refer to https://nt.gov...   
191  Listed as Threatened - refer to https://nt.gov...   

                           new_url new_geoprivacy new_place_id new_username  \
135  https://nt.gov.au/environment       obscured         9994    peggydnew   
185  https://nt.gov.au/environment       obscured         9994    peggydnew   
4    https://nt.gov.au/environment       obscured         9994    peggydnew   
136  https://nt.gov.au/environment       obscured         9994    peggydnew   
5    https://nt.gov.au/environment       obscured         9994    peggydnew   
..                           

## Updates

In [62]:
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username,description
updates = pd.DataFrame(mergedstatuses[mergedstatuses['status_id'].notnull()])
updates.sort_values('scientificName')
updates['action'] = 'UPDATE'
#updates.loc[:,'action'] = 'UPDATE'
updates = updates[['action','scientificName','status_id','taxon_id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
updates.columns = updates.columns.str.replace("new_", "", regex=True)
updates = updates.rename(columns={'scientificName':'taxon_name',
                                  'status_id':'id'})
updates

action              taxon_name      id taxon_id      status  \
159  UPDATE       Bellatorias obiri  152433    38633  Endangered   
57   UPDATE  Hipposideros inornatus  152434    74425  Vulnerable   
176  UPDATE  Pezoporus occidentalis  152431    73180  Endangered   
104  UPDATE    Polytelis alexandrae  152435    19250  Vulnerable   

    iucn_equivalent                                          authority  \
159      Vulnerable   Territory Parks and Wildlife Conservation Act...   
57       Vulnerable   Territory Parks and Wildlife Conservation Act...   
176      Vulnerable   Territory Parks and Wildlife Conservation Act...   
104      Vulnerable   Territory Parks and Wildlife Conservation Act...   

                               url geoprivacy place_id   username  \
159  https://nt.gov.au/environment   obscured     9994  peggydnew   
57   https://nt.gov.au/environment   obscured     9994  peggydnew   
176  https://nt.gov.au/environment   obscured     9994  peggydnew   
104  https://nt.gov.au/environment   obscured     9994  peggydnew   

                                           description  
159  Listed as Threatened - refer to https://nt.gov...  
57   Listed as Threatened - refer to https://nt.gov...  
176  Listed as Threatened - refer to https://nt.gov...  
104  Listed as Threatened - refer to https://nt.gov...

## No status in iNaturalist via straight scientificName match
The NT records that didn't match up to a status in iNaturalist

In [63]:
# to add: those that have no inaturalist status - 532!!
noinatstatus = mergedstatuses[mergedstatuses['status_id'].isnull()]
# try to match the taxon name to something in inaturalist
noinatstatus = noinatstatus.merge(inattaxa, how="left", left_on="scientificName",right_on="scientificName")
noinatstatus

scientificName                 status bionet_geoprivacy status_id  \
0    Abrodictyum obscurum             Endangered              open       NaN   
1    Acacia equisetifolia  Critically Endangered              open       NaN   
2           Acacia latzii             Vulnerable              open       NaN   
3            Acacia peuce             Endangered              open       NaN   
4     Acacia praetermissa             Vulnerable              open       NaN   
..                    ...                    ...               ...       ...   
195    Xylopia monosperma             Endangered              open       NaN   
196       Zeuxine oblonga             Vulnerable              open       NaN   
197         Zyzomys maini             Vulnerable              open       NaN   
198     Zyzomys palatalis             Endangered              open       NaN   
199  Zyzomys pedunculatus  Critically Endangered              open       NaN   

    taxon_id user_id description iucn authority status_inat  ...  \
0        NaN     NaN         NaN  NaN       NaN         NaN  ...   
1        NaN     NaN         NaN  NaN       NaN         NaN  ...   
2        NaN     NaN         NaN  NaN       NaN         NaN  ...   
3        NaN     NaN         NaN  NaN       NaN         NaN  ...   
4        NaN     NaN         NaN  NaN       NaN         NaN  ...   
..       ...     ...         ...  ...       ...         ...  ...   
195      NaN     NaN         NaN  NaN       NaN         NaN  ...   
196      NaN     NaN         NaN  NaN       NaN         NaN  ...   
197      NaN     NaN         NaN  NaN       NaN         NaN  ...   
198      NaN     NaN         NaN  NaN       NaN         NaN  ...   
199      NaN     NaN         NaN  NaN       NaN         NaN  ...   

           phylum          class        order       family    genus  \
0             NaN            NaN          NaN          NaN      NaN   
1    Tracheophyta  Magnoliopsida      Fabales     Fabaceae   Acacia   
2    Tracheophyta  Magnoliopsida      Fabales     Fabaceae   Acacia   
3    Tracheophyta  Magnoliopsida      Fabales     Fabaceae   Acacia   
4    Tracheophyta  Magnoliopsida      Fabales     Fabaceae   Acacia   
..            ...            ...          ...          ...      ...   
195           NaN            NaN          NaN          NaN      NaN   
196  Tracheophyta     Liliopsida  Asparagales  Orchidaceae  Zeuxine   
197      Chordata       Mammalia     Rodentia      Muridae  Zyzomys   
198      Chordata       Mammalia     Rodentia      Muridae  Zyzomys   
199      Chordata       Mammalia     Rodentia      Muridae  Zyzomys   

    specificEpithet infraspecificEpithet              modified taxonRank  \
0               NaN                  NaN                   NaN       NaN   
1     equisetifolia                  NaN  2022-04-06T22:05:25Z   species   
2            latzii                  NaN  2022-04-07T02:06:43Z   species   
3             peuce                  NaN  2022-04-06T23:48:13Z   species   
4      praetermissa                  NaN  2022-04-05T02:10:02Z   species   
..              ...                  ...                   ...       ...   
195             NaN                  NaN                   NaN       NaN   
196         oblonga                  NaN  2021-07-16T02:50:44Z   species   
197           maini                  NaN  2019-08-27T01:49:16Z   species   
198       palatalis                  NaN  2019-11-22T22:46:28Z   species   
199    pedunculatus                  NaN  2019-11-22T22:46:29Z   species   

                                            references  
0                                                  NaN  
1                       https://eol.org/pages/49426174  
2    http://www.catalogueoflife.org/annual-checklis...  
3    http://www.catalogueoflife.org/annual-checklis...  
4    http://www.catalogueoflife.org/annual-checklis...  
..                                                 ...  
195                                                NaN  
196  http:/

In [64]:
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions

scientificName                 status bionet_geoprivacy status_id  \
1    Acacia equisetifolia  Critically Endangered              open       NaN   
2           Acacia latzii             Vulnerable              open       NaN   
3            Acacia peuce             Endangered              open       NaN   
4     Acacia praetermissa             Vulnerable              open       NaN   
5       Acacia undoolyana             Vulnerable              open       NaN   
..                    ...                    ...               ...       ...   
193      Vidumelon wattii             Vulnerable              open       NaN   
196       Zeuxine oblonga             Vulnerable              open       NaN   
197         Zyzomys maini             Vulnerable              open       NaN   
198     Zyzomys palatalis             Endangered              open       NaN   
199  Zyzomys pedunculatus  Critically Endangered              open       NaN   

    taxon_id user_id description iucn authority status_inat  ...  \
1        NaN     NaN         NaN  NaN       NaN         NaN  ...   
2        NaN     NaN         NaN  NaN       NaN         NaN  ...   
3        NaN     NaN         NaN  NaN       NaN         NaN  ...   
4        NaN     NaN         NaN  NaN       NaN         NaN  ...   
5        NaN     NaN         NaN  NaN       NaN         NaN  ...   
..       ...     ...         ...  ...       ...         ...  ...   
193      NaN     NaN         NaN  NaN       NaN         NaN  ...   
196      NaN     NaN         NaN  NaN       NaN         NaN  ...   
197      NaN     NaN         NaN  NaN       NaN         NaN  ...   
198      NaN     NaN         NaN  NaN       NaN         NaN  ...   
199      NaN     NaN         NaN  NaN       NaN         NaN  ...   

           phylum          class            order       family      genus  \
1    Tracheophyta  Magnoliopsida          Fabales     Fabaceae     Acacia   
2    Tracheophyta  Magnoliopsida          Fabales     Fabaceae     Acacia   
3    Tracheophyta  Magnoliopsida          Fabales     Fabaceae     Acacia   
4    Tracheophyta  Magnoliopsida          Fabales     Fabaceae     Acacia   
5    Tracheophyta  Magnoliopsida          Fabales     Fabaceae     Acacia   
..            ...            ...              ...          ...        ...   
193      Mollusca     Gastropoda  Stylommatophora   Camaenidae  Vidumelon   
196  Tracheophyta     Liliopsida      Asparagales  Orchidaceae    Zeuxine   
197      Chordata       Mammalia         Rodentia      Muridae    Zyzomys   
198      Chordata       Mammalia         Rodentia      Muridae    Zyzomys   
199      Chordata       Mammalia         Rodentia      Muridae    Zyzomys   

    specificEpithet infraspecificEpithet              modified taxonRank  \
1     equisetifolia                  NaN  2022-04-06T22:05:25Z   species   
2            latzii                  NaN  2022-04-07T02:06:43Z   species   
3             peuce                  NaN  2022-04-06T23:48:13Z   species   
4      praetermissa                  NaN  2022-04-05T02:10:02Z   species   
5        undoolyana                  NaN  2022-04-05T02:34:14Z   species   
..              ...                  ...                   ...       ...   
193          wattii                  NaN  2021-10-29T15:37:54Z   species   
196         oblonga                  NaN  2021-07-16T02:50:44Z   species   
197           maini                  NaN  2019-08-27T01:49:16Z   species   
198       palatalis                  NaN  2019-11-22T22:46:28Z   species   
199    pedunculatus                  NaN  2019-11-22T22:46:29Z   species   

                                            references  
1                       https://eol.org/pages/49426174  
2    http://www.catalogueoflife.org/annual-checklis...  
3    http://www.catalogueoflife.org/annual-checklis...  
4    http://www.catalogueoflife.org/annual-checklis...  
5    http://www.catalogueoflife.org/annual-checklis...  
..                                                 ..

In [65]:
# there's no status but there is a matching inat taxon (id is the taxon id)
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions.sort_values(['scientificName'])
additions['action'] = 'ADD'
additions = additions[['action','scientificName','status_id','id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
additions.columns = additions.columns.str.replace("new_", "", regex=True)
additions = additions.rename(columns={'scientificName':'taxon_name',
                                      'id':'taxon_id',
                                  'status_id':'id'})
additions

action            taxon_name   id taxon_id                 status  \
1      ADD  Acacia equisetifolia  NaN  1253756  Critically Endangered   
2      ADD         Acacia latzii  NaN  1254327             Vulnerable   
3      ADD          Acacia peuce  NaN   465191             Endangered   
4      ADD   Acacia praetermissa  NaN  1254561             Vulnerable   
5      ADD     Acacia undoolyana  NaN  1254884             Vulnerable   
..     ...                   ...  ...      ...                    ...   
193    ADD      Vidumelon wattii  NaN   114966             Vulnerable   
196    ADD       Zeuxine oblonga  NaN   369267             Vulnerable   
197    ADD         Zyzomys maini  NaN    45377             Vulnerable   
198    ADD     Zyzomys palatalis  NaN    75238             Endangered   
199    ADD  Zyzomys pedunculatus  NaN    45374  Critically Endangered   

    iucn_equivalent                                          authority  \
1        Vulnerable   Territory Parks and Wildlife Conservation Act...   
2        Vulnerable   Territory Parks and Wildlife Conservation Act...   
3        Vulnerable   Territory Parks and Wildlife Conservation Act...   
4        Vulnerable   Territory Parks and Wildlife Conservation Act...   
5        Vulnerable   Territory Parks and Wildlife Conservation Act...   
..              ...                                                ...   
193      Vulnerable   Territory Parks and Wildlife Conservation Act...   
196      Vulnerable   Territory Parks and Wildlife Conservation Act...   
197      Vulnerable   Territory Parks and Wildlife Conservation Act...   
198      Vulnerable   Territory Parks and Wildlife Conservation Act...   
199      Vulnerable   Territory Parks and Wildlife Conservation Act...   

                               url geoprivacy place_id   username  \
1    https://nt.gov.au/environment   obscured     9994  peggydnew   
2    https://nt.gov.au/environment   obscured     9994  peggydnew   
3    https://nt.gov.au/environment   obscured     9994  peggydnew   
4    https://nt.gov.au/environment   obscured     9994  peggydnew   
5    https://nt.gov.au/environment   obscured     9994  peggydnew   
..                             ...        ...      ...        ...   
193  https://nt.gov.au/environment   obscured     9994  peggydnew   
196  https://nt.gov.au/environment   obscured     9994  peggydnew   
197  https://nt.gov.au/environment   obscured     9994  peggydnew   
198  https://nt.gov.au/environment   obscured     9994  peggydnew   
199  https://nt.gov.au/environment   obscured     9994  peggydnew   

                                           description  
1    Listed as Threatened - refer to https://nt.gov...  
2    Listed as Threatened - refer to https://nt.gov...  
3    Listed as Threatened - refer to https://nt.gov...  
4    Listed as Threatened - refer to https://nt.gov...  
5    Listed as Threatened - refer to https://nt.gov...  
..                                                 ...  
193  Listed as Threatened - refer to https://nt.gov...  
196  Listed as Threatened - refer to https://nt.gov...  
197  Listed as Threatened - refer to https://nt.gov...  
198  Listed as Threatened - refer to https://nt.gov...  
199  Listed as Threatened - refer to https://nt.gov...  

[136 rows x 12 columns]

In [66]:
# write these to the file
pd.concat([updates,additions]).to_csv(sourcedir + "nt.csv", index=False)

In [67]:
# what didnt match to a taxon?
unknownToInat = noinatstatus[noinatstatus['id'].isna()]
unknownToInat

scientificName                 status bionet_geoprivacy  \
0      Abrodictyum obscurum             Endangered              open   
6       Acanthophsis hawkei             Vulnerable              open   
9           Amperia spicata             Vulnerable              open   
19       Babingtonia behrii             Vulnerable              open   
20      Baumea arthrophylla             Endangered              open   
..                      ...                    ...               ...   
181      Typhonium mirabile             Endangered              open   
183       Typhonium taylori             Endangered              open   
184  Typhonium sp. Sandover             Vulnerable              open   
194  Vincentrachia desmonda  Critically Endangered              open   
195      Xylopia monosperma             Endangered              open   

    status_id taxon_id user_id description iucn authority status_inat  ...  \
0         NaN      NaN     NaN         NaN  NaN       NaN         NaN  ...   
6         NaN      NaN     NaN         NaN  NaN       NaN         NaN  ...   
9         NaN      NaN     NaN         NaN  NaN       NaN         NaN  ...   
19        NaN      NaN     NaN         NaN  NaN       NaN         NaN  ...   
20        NaN      NaN     NaN         NaN  NaN       NaN         NaN  ...   
..        ...      ...     ...         ...  ...       ...         ...  ...   
181       NaN      NaN     NaN         NaN  NaN       NaN         NaN  ...   
183       NaN      NaN     NaN         NaN  NaN       NaN         NaN  ...   
184       NaN      NaN     NaN         NaN  NaN       NaN         NaN  ...   
194       NaN      NaN     NaN         NaN  NaN       NaN         NaN  ...   
195       NaN      NaN     NaN         NaN  NaN       NaN         NaN  ...   

    phylum class order family genus specificEpithet infraspecificEpithet  \
0      NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
6      NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
9      NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
19     NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
20     NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
..     ...   ...   ...    ...   ...             ...                  ...   
181    NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
183    NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
184    NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
194    NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
195    NaN   NaN   NaN    NaN   NaN             NaN                  NaN   

    modified taxonRank references  
0        NaN       NaN        NaN  
6        NaN       NaN        NaN  
9        NaN       NaN        NaN  
19       NaN       NaN        NaN  
20       NaN       NaN        NaN  
..       ...       ...        ...  
181      NaN       NaN        NaN  
183      NaN       NaN        NaN  
184      NaN       NaN        NaN  
194      NaN       NaN        NaN  
195      NaN       NaN        NaN  

[64 rows x 36 columns]

### are there any that need to be removed?
NT sensitive list count: 0
NT inat statuses count: 12

updates to inat status: 4
additional inat status: 136
NT statuses we can't find a taxon match for in iNaturalist: 64
total: 166 (explainable via the various genus/section entries that we matched to in the taxonomy)

inat statuses left over: 12-4=8 that may need checking

In [68]:
# inat statuses that aren't in added or updated
inatstatuses[~inatstatuses['taxon_id'].isin(updates['taxon_id'])]


status_id taxon_id user_id place_id source_id                  authority  \
1001    162724  1134239  702203     9994                   Northern Territory   
927     165448    12647  702203     9994                   Northern Territory   
289     263901  1289379  702203     9994            Atlas of Living Australia   
215     170163    20166  702203     9994            Atlas of Living Australia   
1252    152436    40743  708886     9994     16652                  NT NRETAS   
1248    152432    41326  708886     9994     16652                  NT NRETAS   
986     139906   698942             9994                                        
1       234788   918383  702203     9994            Atlas of Living Australia   

             status                                                url  \
1001             LC  https://bie.ala.org.au/species/https://id.biod...   
927              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
289              VU  https://bie.ala.org.au/species/urn:lsid:biodiv...   
215              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
1252     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1248     endangered  https://lists.ala.org.au/speciesListItem/list/...   
986   least concern  https://bie.ala.org.au/species/http://id.biodi...   
1               NT   https://bie.ala.org.au/species/https://id.biod...   

                          description geoprivacy  ...         genus  \
1001                                        open  ...    Cryptandra   
927                                         open  ...    Epthianura   
289                                         open  ...      Chloebia   
215                                               ...         Ninox   
1252                                    obscured  ...                 
1248                                    obscured  ...    Macroderma   
986   Atlas of Living Australia (ALA)             ...      Duboisia   
1                                                 ...  Chiloschista   

     specificEpithet infraspecificEpithet              modified  \
1001         gemmata                       2020-09-25T18:01:37Z   
927           crocea                       2021-09-17T08:46:17Z   
289         gouldiae                       2022-10-20T02:55:37Z   
215        connivens                       2021-07-28T02:17:03Z   
1252                                                              
1248           gigas                       2019-08-27T01:58:05Z   
986        hopwoodii                       2019-02-16T13:19:18Z   
1        phyllorhiza                       2022-01-08T03:30:36Z   

                scientificName taxonRank  \
1001        Cryptandra gemmata   species   
927          Epthianura crocea   species   
289          Chloebia gouldiae   species   
215            Ninox connivens   species   
1252     Hipposideros stenotis             
1248          Macroderma gigas   species   
986         Duboisia hopwoodii   species   
1     Chiloschista phyllorhiza   species   

                                             references  \
1001  http://plantsoftheworldonline.org/taxon/urn:ls...   
927   http://www.birdlife.org/datazone/speciesfactsh...   
289   https://www.birds.cornell.edu/clementschecklis...   
215   http://www.birdlife.org/datazone/speciesfactsh...   
1252                                                      
1248  http://www.catalogueoflife.org/annual-checklis...   
986                             https://www.ala.org.au/   
1     http://www.catalogueoflife.org/annual-checklis...   

           preferred_common_name is_active current_synonymous_taxon_ids  
1001                                                                     
927                                                                      
289                                                                      
215                                                                      
1252  Narrow-eared Roundleaf Bat     False                    [14